### DISCUSSION TITLE
Allow mathmatical operators on record creation in with-block

### DISCUSSION THREAD

**Discussion by Khaos66** [dotnet/csharplang#5040] [Language Ideas]
# Idea
Let us use operators inside the `with` block.

# Example
```C# 
record MyRecord(int Number);

var instance1 = new MyRecord(10);
var instance2 = instance1 with { Number += 5 };
```

Should be the same as
```C#
...
var instance2 = instance1 with { Number = instance1.Number + 5 };
```

**Comment by indiemini** [comment]
I wanted to echo that this would make the `with` operator *extremely* convenient to use. Since the original proposal hasn't seen much activity, I thought I would try expand on it a bit (and avoid opening a duplicate discussion).

# Proposal
**Compound assignment operators within a `with` initialization should be be valid.**

# Semantics

```cs
using System;

public class CompoundAssignmentExample
{
    public record AbilityInfo(float cooldown, float damage);

    public AbilityInfo ApplyCharacterStats(Character c, AbilityInfo ability)
    {
        // get character's base stats
        float dmgMod = c.GetDmgModifier();
        float cdr = c.GetCooldownReduction();
        // use compound assignment operators to create the updated value
        ability = ability with { cooldown *= cdr, damage *= dmgMod };
        return ability;
    }

    // Silly example class
    public class Character
    {
        public float GetDmgModifier() => 1f;
        public float GetCooldownReduction() => 1f;
    }
}
```

# Goal
When I use the `with` operator, it's always with the goal of "write less, do more." The `with` operator reduces the amount of syntax required to update a `record` or `struct` value non-destructively--but if you want to do the same thing to one of its properties, it shouldn't take more text.

This syntax is easily understood, and currently isn't valid, so there shouldn't be any issues with ambiguity in existing code. This proposal is a really just a small quality-of-life improvement--but then again, ain't that the whole `with` operator? :wink:

**Comment by jnm2** [comment]
Existing discussions:

- https://github.com/dotnet/csharplang/discussions/4494
- https://github.com/dotnet/csharplang/issues/5176#issuecomment-918822909 where @CyrusNajmabadi showed interest in this
